File: 28-implement-strstr/28-implement-strstr.cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        //using kmp algorithm.
        int n = haystack.length(), m = needle.length();
        
        if(needle.length() == 0)
            return 0;
        
        vector<int> lps(m, 0);
        int prev = 0, curr = 1;
        
        //fill lps array.
        while(curr < m) {
            if(needle[curr] == needle[prev])
                lps[curr++] = prev++ + 1;
            else {
                if(prev == 0)
                    lps[curr++] = 0;
                else
                    prev = lps[prev - 1];
            }
        }
        
        int i = 0, j = 0;
        while(i < n) {
            if(haystack[i] == needle[j])
                i++, j++;
            
            else {
                if(j == 0)
                    i++;
                else
                    j = lps[j - 1];
            }
            
            if(j == m)
                return i - m;
        }
        
        return -1;
    }
};

File: 28-implement-strstr/28-implement-strstr.java
class Solution {
    public int strStr(String haystack, String needle) {
        //brute
        int len = haystack.length();
        int j = 0;
        int i;
        for(i=0;i<len;i++)
        {
            if(haystack.charAt(i) == needle.charAt(j))
                j++;
            else
            {
                i = i - j; // return to index where we start matching.
                j = 0;
            }
            if(j == needle.length())
                return i -j + 1;
        }
        return -1;
    }
}